# Investigating DNA Words and Their Distributions Across the Tree of Life

## Î‘bstract
The frequency distributions of DNA k-mers are shaped by fundamental biological processes and offer a window into genome structure and evolution. Inspired by analogies to natural language, prior studies have attempted to model genomic k-mer usage using Zipfâ€™s law, a rank-frequency law originally formulated for words in human language. However, the extent to which this law accurately captures the distribution of k-mers across diverse species remains unclear. Here, we systematically analyze k-mer frequency spectra across more than 225,000 genome assemblies spanning all three domains of life and viruses. We demonstrate that Zipfâ€™s law consistently underperforms in modeling k-mer distributions. In contrast, we propose the truncated power law and Zipf-Mandelbrot distributions, which provide substantially improved fits across taxonomic groups. We show that genome size and GC content influence model performance, with larger and GC-content imbalanced genomes yielding better adherence. Additionally, we perform an extensive analysis on vocabulary expansion and exhaustion across the same organisms using Heapsâ€™ law. We apply our modeling framework to evaluate simulated genomes generated by k-let preserving shuffling and deep generative language models. Our results reveal substantial differences between organismal genomes and their synthetic or shuffled counterparts, offering a novel approach to benchmark the biological plausibility of artificial genomes. Collectively, this work establishes new standards for modeling genomic k-mer distributions and provides insights relevant to synthetic biology, and evolutionary sequence analysis.

## ðŸ“‚ Directory Overview

- **`artificial_genomes/`**  
  Contains the scripts for the creation and fitting of shuffled and synthetic genomes,  
  as well as the figures from Figure 7.

- **`determinants/`**  
  Analysis on the factors that determine the goodness of fit of the Zipf-Mandelbrot and the truncated power law.  
  The subfolder `genome_size` has all the figures relevant to genome size (Figure 6 and Supplementary Figure 6).  
  `gc_content` relates to Figure 6 and Supplementary Figure 7, and `genic_percentage` to Figure 6 and Supplementary Figure 8.

- **`global_patterns/`**  
  Refers to the creation of the mean counts vs ranks plots (Figure 1),  
  the Lorentz curves (Figure 1, Supplementary Figure 1),  
  and the Gini coefficients (Figure 1, Supplementary Figure 2).

- **`heaps_law/`**  
  K-mer vocabulary-growth experiments using Heap's Law. Represents Figures 2 and 3.

- **`model_fits/`**  
  Contains the main results for the fit of the Zipf-Mandelbrot and the truncated power law to 225k complete genomes.  
  Particularly, `Results_truncated.xlsx` and `Results_Zipf_Mandelbrot.xlsx` contain information about the original genome  
  (such as accession, assembly, and genome size), as well as the fitted parameters for each of the two distributions  
  for k-mer lengths 3, 4, 5, 6, and 7. Corresponds to Figure 4 and Supplementary Figures 4 and 5.

- **`not_zipf/`**  
  Analysis on why genomes deviate from a plain Zipf's Law.  
  Contains figures from Figure 4 and Supplementary Figure 3.


## ðŸ§° Libraries Used

â€¢ [NumPy](https://numpy.org/) â€¢ [Biopython](https://biopython.org/) â€¢ [SciPy](https://scipy.org/) â€¢ [Matplotlib](https://matplotlib.org/) â€¢ [Seaborn](https://seaborn.pydata.org/) â€¢ [Pandas](https://pandas.pydata.org/)